src/api.jl

```diff
@@ -59,5 +59,95 @@ function knitro_statuses(code::Integer)
   return :unknown
 end
 
+mutable struct KnitroSolver
+  kc
+end
+
+function Base.finalize(solver::KnitroSolver)
+  KNITRO.KN_reset_params_to_defaults(solver.kc)
+  KNITRO.KN_free(solver.kc)
+end
+
+# pass options to KNITRO
+function setparams!(solver::KnitroSolver; kwargs...)
+  kc = solver.kc
+  KNITRO.KN_reset_params_to_defaults(kc)
+  # set primal and dual initial guess
+  kwargs = Dict(kwargs)
+  if :x0 ∈ keys(kwargs)
+    KNITRO.KN_set_var_primal_init_values(kc, kwargs[:x0])
+    pop!(kwargs, :x0)
+  end
+  if :y0 ∈ keys(kwargs)
+    KNITRO.KN_set_con_dual_init_values(kc, kwargs[:y0])
+    pop!(kwargs, :y0)
+  end
+  if :z0 ∈ keys(kwargs)
+    KNITRO.KN_set_var_dual_init_values(kc, kwargs[:z0])
+    pop!(kwargs, :z0)
+  end
+
+  # specify that we are able to provide the Hessian without including the objective
+  KNITRO.KN_set_param(kc, KNITRO.KN_PARAM_HESSIAN_NO_F, KNITRO.KN_HESSIAN_NO_F_ALLOW)
+
+  # pass options to KNITRO
+  for (k, v) in kwargs
+    KNITRO.KN_set_param(kc, string(k), v)
+  end
+
+  return solver
+end
+
+KnitroSolver(nlp::AbstractNLPModel; kwargs...) = KnitroSolver(_is_general_nlp(nlp), nlp; kwargs...)
+
 include("nlp.jl")
 include("nls.jl")
+
+knitro(nlp::AbstractNLPModel, solver::KnitroSolver) = _knitro(nlp, solver)
+
+function _knitro(
+  val,
+  nlp::AbstractNLPModel;
+  callback::Union{Function, Nothing} = nothing,
+  kwargs...,
+)
+  solver = KnitroSolver(val, nlp, callback = callback; kwargs...)
+  stats = _knitro(nlp, solver)
+  finalize(solver)
+  return stats
+end
+
+function _knitro(nlp::AbstractNLPModel, solver::KnitroSolver)
```

Comment:
This should probably be `_knitro!` now?!

Reply:
you are right, thanks.